skimage/registration/_optical_flow.py

```diff
@@ -217,3 +218,152 @@ def optical_flow_tvl1(reference_image, moving_image,
                      tol=tol, prefilter=prefilter)
 
     return coarse_to_fine(reference_image, moving_image, solver, dtype=dtype)
+
+
+def _ilk(reference_image, moving_image, flow0, radius, num_warp, gaussian,
+         prefilter):
+    """Iterative Lucas-Kanade (iLK) solver for optical flow estimation.
+
+    Parameters
+    ----------
+    reference_image : ndarray, shape (M, N[, P[, ...]])
+        The first gray scale image of the sequence.
+    moving_image : ndarray, shape (M, N[, P[, ...]])
+        The second gray scale image of the sequence.
+    flow0 : ndarray, shape (reference_image.ndim, M, N[, P[, ...]])
+        Initialization for the vector field.
+    radius : int
+        Radius of the window considered around each pixel.
+    num_warp : int
+        Number of times moving_image is warped.
+    gaussian : bool
+        if True, a gaussian kernel is used for the local
+        integration. Otherwise, a uniform kernel is used.
+    prefilter : bool
+        Whether to prefilter the estimated optical flow before each
+        image warp. This helps to remove potential outliers.
+
+    Returns
+    -------
+    flow : ndarray, shape ((reference_image.ndim, M, N[, P[, ...]])
+        The estimated optical flow components for each axis.
+
+    """
+    dtype = reference_image.dtype
+    ndim = reference_image.ndim
+    size = 2 * radius + 1
+
+    if gaussian:
+        sigma = (0, ) + ndim * (size / 4, )
+        filter_func = partial(ndi.gaussian_filter, sigma=sigma, mode='mirror')
+    else:
+        filter_func = partial(ndi.uniform_filter, size=ndim * (size, ),
+                              mode='mirror')
+
+    flow = flow0
+    # For each pixel location (i, j), the optical flow X = flow[:, i, j]
+    # is the solution of the ndim x ndim linear system
+    # A[i, j] * X = b[i, j]
+    A = np.zeros(reference_image.shape + (ndim, ndim), dtype=dtype)
+    b = np.zeros(reference_image.shape + (ndim, ), dtype=dtype)
+
+    grid = np.meshgrid(*[np.arange(n, dtype=dtype)
+                         for n in reference_image.shape],
+                       indexing='ij', sparse=True)
+
+    for _ in range(num_warp):
+        if prefilter:
+            flow = ndi.filters.median_filter(flow, (1, ) + ndim * (3, ))
+
+        moving_image_warp = warp(moving_image, get_warp_points(grid, flow),
+                                 mode='nearest')
+        grad = np.array(np.gradient(moving_image_warp))
+        It = (grad * flow).sum(axis=0) + reference_image - moving_image_warp
+
+        # Local linear systems creation
+        for i, j in combinations_with_replacement(range(ndim), 2):
+            A[..., i, j] = A[..., j, i] = filter_func(grad[i] * grad[j])
```

Comment:
If I am reading the referenced Paley2016 publication section 2.2 correctly, this computation as well as the `grad` computation should occur outside of this `num_warps` loop.
They mention that the Hessians, `A` (**H** in the publication), are computed "during initialization". Specifically, the following should be placed prior to the `num_warps` loop and then the `grad` and `A` computations removed from the loop. (Note also the change to "reference_image" as the input to the gradient computation)

```Python
    grad = np.array(np.gradient(reference_image))
    # Local linear systems creation
    for i, j in combinations_with_replacement(range(ndim), 2):
        A[..., i, j] = A[..., j, i] = filter_func(grad[i] * grad[j])
```


So, the loop will only update `flow`, `b` and `moving_image_warp` warp at each iteration, but not `grad` or `A`.

The test suite still passes for me with this change.

Reply:
@grlee77, you are perfectly reading section 2.2 of [[2]](https://www.researchgate.net/profile/Aurelien_Plyer/publication/262948097_Massively_Parallel_Lucas_Kanade_Optical_Flow_for_Real-Time_Video_Processing_Applications/links/54b4db3e0cf2318f0f96f01c.pdf) :wink:, but the algorithm described there solves the *inverse* form fo Lucas-Kanade algo (see section 3.2 [here](http://ncorr.com/download/publications/bakerunify.pdf) as ref for example). This form is briefly discussed in section 3.2 of [[1]](https://www.researchgate.net/profile/Frederic_Champagnat/publication/4186265_Dense_optical_flow_by_iterative_local_window_registration/links/5d25b1ef92851cf44074e8c0/Dense-optical-flow-by-iterative-local-window-registration.pdf).
Current implementation is described in Table 2 of [1].

Reply:
Okay, thanks. So, this "MIWS" version is a little slower, but should have better stability. I just left a new comment about adding a Note in the docstring guiding the user interested in details to the specific algorithm within the publications that was used.